### PR TITLE
Issue#67 rest controller for search

### PR DESCRIPTION
**Pull request by hedgehog1833**
There is a new Rest Endpoint for searching artists. The old MVC Controller just returns the search field, the rest is done via Javascript now.
The tests are not ideal yet, I know that. I encountered some problems while testing Rest and MVC, we should talk about that separately I think.

### CODE REVIEW

File: src/main/java/com/metalr2/config/constants/Endpoints.java [hunk 1]
```diff
@@ -44,7 +44,8 @@
   }
 
   public static class Rest {
-    public static final String FOLLOW_ARTISTS_V1     = "/rest/v1/follow-artist";
+    public static final String FOLLOW_ARTISTS_V1 = "/rest/v1/follow-artist";
+    public static final String ARTISTS           = "/rest/v1/artist";
```
Comment (Waginator) [inline src/main/java/com/metalr2/config/constants/Endpoints.java]: Naming convention of fields is different...

File: src/main/java/com/metalr2/web/controller/ArtistDetailsController.java [hunk 2]
```diff
@@ -30,15 +29,15 @@
 @Slf4j
 public class ArtistDetailsController {
 
-  private final ArtistSearchRestClient artistSearchRestClient;
+  private final DiscogsArtistSearchRestClientImpl discogsArtistSearchRestClient;
```
Comment (Waginator) [inline src/main/java/com/metalr2/web/controller/ArtistDetailsController.java]: - Please use the interface as type
- maybe artistSearchClient is a little bit smaller and also expressive?

File: src/main/java/com/metalr2/web/dto/discogs/artist/DiscogsArtist.java [hunk 3]
```diff
@@ -20,7 +20,7 @@
         "members"
 })
 @ToString(of = {"id"})
-public class Artist {
+public class DiscogsArtist {
```
Comment (Waginator) [inline src/main/java/com/metalr2/web/dto/discogs/artist/DiscogsArtist.java]: Thumbs up for the renamings!

File: src/test/java/com/metalr2/web/controller/SearchArtistsControllerIT.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    });
+
+    // Login
+    MvcResult result = mockMvc.perform(post(Endpoints.Guest.LOGIN)
+            .param("username", USERNAME)
+            .param("password", PASSWORD))
+            .andReturn();
+
+    session = (MockHttpSession)result.getRequest().getSession();
+  }
```
Comment (Waginator) [inline src/test/java/com/metalr2/web/controller/SearchArtistsControllerIT.java]: Just a reminder that test should be executed within an "authentication-less" mode.

File: src/test/java/com/metalr2/web/controller/SearchArtistRestControllerIT.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    String requestUri   = "http://" + serverAddress + ":" + port + Endpoints.Rest.ARTISTS;
+    requestHandler      = new RestAssuredRequestHandler<>(requestUri, USERNAME, PASSWORD);
+    userRepository.save(UserEntity.builder()
+            .username(USERNAME)
+            .email(EMAIL)
+            .password("$2a$10$2IevDskxEeSmy7Sy41Xl7.u22hTcw3saxQghS.bWaIx3NQrzKTvxK")
+            .enabled(true)
+            .userRoles(UserRole.createUserRole())
+            .build());
+  }
```
Comment (Waginator) [inline src/test/java/com/metalr2/web/controller/SearchArtistRestControllerIT.java]: Just a reminder that test should be executed within an "authentication-less" mode.

File: src/test/java/com/metalr2/web/controller/SearchArtistRestControllerIT.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    ValidatableResponse validatableResponse = requestHandler.doPost(ContentType.JSON, artistSearchRequest);
+
+    // assert
+    validatableResponse.contentType(ContentType.JSON)
+            .statusCode(HttpStatus.OK.value());
+
+    ArtistNameSearchResponse artistNameSearchResponse = validatableResponse.extract().as(ArtistNameSearchResponse.class);
+
+    assertThat(artistNameSearchResponse).isNotNull();
+    assertThat(artistNameSearchResponse.getPagination()).isNotNull();
```
Comment (Waginator) [inline src/test/java/com/metalr2/web/controller/SearchArtistRestControllerIT.java]: strictly speaking, the not null asserts do not really have any added value...the test will fail in a later assert.

File: src/test/java/com/metalr2/web/controller/SearchArtistRestControllerIT.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    assertThat(artistSearchResult).isNotNull();
+    assertThat(artistSearchResult.getArtistName()).isEqualTo(VALID_SEARCH_REQUEST);
+    assertThat(artistSearchResult.getId()).isEqualTo(DISCOGS_ARTIST_ID);
+
+    Pagination pagination = artistNameSearchResponse.getPagination();
+
+    assertThat(pagination.getCurrentPage()).isEqualTo(DEFAULT_PAGE);
+    assertThat(pagination.getSize()).isEqualTo(DEFAULT_SIZE);
+    assertThat(pagination.getTotalPages()).isEqualTo(DEFAULT_PAGE+1);
+    assertThat(pagination.getNextPage()).isEqualTo(DEFAULT_PAGE+1);
```
Comment (Waginator) [inline src/test/java/com/metalr2/web/controller/SearchArtistRestControllerIT.java]: You can also assert against a whole object...

`assertThat(pagination).isEqualTo(new Pagination(DEFAULT_PAGE, DEFAULT_SIZE, DEFAULT_PAGE + 1, DEFAULT_PAGE + 1))`

Check also other asserts...

And maybe you should define a fixed value for the total page assert. The value is not dependent on the current page, isn't it?

File: src/test/java/com/metalr2/web/controller/FollowArtistRestControllerIT.java [hunk 8]
```diff
@@ -52,16 +53,17 @@
...
+    followArtistRequest = new FollowArtistRequest(ARTIST_NAME, ARTIST_DISCOGS_ID);
+    userEntity          = UserEntity.builder()
+            .username(USERNAME)
+            .email(EMAIL)
+            .password("$2a$10$2IevDskxEeSmy7Sy41Xl7.u22hTcw3saxQghS.bWaIx3NQrzKTvxK")
+            .enabled(true)
+            .userRoles(UserRole.createUserRole())
+            .build();
+    userRepository.save(userEntity);
+    followArtistDto     = new FollowArtistDto(userEntity.getPublicId(), ARTIST_NAME, ARTIST_DISCOGS_ID);
```
Comment (Waginator) [inline src/test/java/com/metalr2/web/controller/FollowArtistRestControllerIT.java]: Just a reminder that test should be executed within an "authentication-less" mode.

File: src/main/java/com/metalr2/web/controller/SearchArtistRestController.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.stream.Collectors;
+
+@RestController
+@RequestMapping({Endpoints.Rest.ARTISTS})
+public class SearchArtistRestController {
+
+  private static final int DEFAULT_PAGE_SIZE = 25;
+  private static final int DEFAULT_PAGE = 1;
+
+  private final DiscogsArtistSearchRestClient discogsArtistSearchRestClient;
```
Comment (Waginator) [inline src/main/java/com/metalr2/web/controller/SearchArtistRestController.java]: maybe `artistSearchClient`

File: src/main/java/com/metalr2/web/controller/SearchArtistRestController.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    Optional<DiscogsArtistSearchResultContainer> artistSearchResultsOptional = discogsArtistSearchRestClient.searchByName(artistSearchRequest.getArtistName(),
+            artistSearchRequest.getPage(), artistSearchRequest.getSize());
+
+    if (artistSearchResultsOptional.isEmpty()) {
+      return new ArtistNameSearchResponse(Collections.emptyList(),new Pagination());
+    }
+
+    DiscogsArtistSearchResultContainer discogsArtistSearchResults = artistSearchResultsOptional.get();
+
+    return createArtistNameSearchResponse(discogsArtistSearchResults, publicUserId);
```
Comment (Waginator) [inline src/main/java/com/metalr2/web/controller/SearchArtistRestController.java]: `return createArtistNameSearchResponse(artistSearchResultsOptional.get(), publicUserId);`

File: src/main/java/com/metalr2/web/controller/SearchArtistRestController.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+  private ArtistNameSearchResponse createArtistNameSearchResponse(DiscogsArtistSearchResultContainer artistSearchResults, String publicUserId) {
+    DiscogsPagination discogsPagination         = artistSearchResults.getDiscogsPagination();
+    DiscogsPaginationUrls discogsPaginationUrls = discogsPagination.getUrls();
+
+    int size      = discogsPaginationUrls.getNext() != null ? discogsPagination.getItemsPerPage() : DEFAULT_PAGE_SIZE;
+    int nextPage  = discogsPaginationUrls.getNext() != null ? discogsPagination.getCurrentPage() + 1 : DEFAULT_PAGE;
+
+    List<FollowArtistDto> alreadyFollowedArtists = followArtistService.findPerUser(publicUserId);
+
+    List<ArtistNameSearchResponse.ArtistSearchResult> dtoArtistSearchResults = artistSearchResults.getResults().stream()
```
Comment (Waginator) [inline src/main/java/com/metalr2/web/controller/SearchArtistRestController.java]: You can do a static import for `ArtistSearchResult`. `ArtistNameSearchResponse` can be left out.

File: src/main/java/com/metalr2/web/controller/SearchArtistRestController.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+            .map(artistSearchResult -> new ArtistNameSearchResponse.ArtistSearchResult(artistSearchResult.getThumb(),
+                    artistSearchResult.getId(), artistSearchResult.getTitle(),
+                    alreadyFollowedArtists.contains(new FollowArtistDto(publicUserId, artistSearchResult.getTitle(), artistSearchResult.getId()))))
```
Comment (Waginator) [inline src/main/java/com/metalr2/web/controller/SearchArtistRestController.java]: We also have quite a big overhead here. What do you think of the following idea:

`List<FollowArtistDto> alreadyFollowedArtists = followArtistService.findPerUser(publicUserId) // ToDo: Collect the Artist IDs`

`alreadyFollowedArtists.contains(artistSearchResult.getId())`

**Comment by hedgehog1833** [comment]
Closes #67

**Pull request merged**